### DISCUSSION TITLE
How to recover ( add it again )  the removed rectangle?

### DISCUSSION THREAD

**Discussion by alamenai** [fabricjs/fabric.js#9760] [Q&A] [answered]
Hi,

I'm trying to simulate ( prototype ) the roof of a house and on top of it, there are several solar panels.

The idea is to allow the user unselect the existing panel and select it again.

I successfully created the canvas ( roof )  and added to it a specific number of rectangles (panels) with specific options.

Also, I'm able to deselect the rectangle ( panel ) but I could not recover it when I select on its space.

#### Initial View

![image](https://github.com/fabricjs/fabric.js/assets/20143684/840ff23d-0de9-4a38-9201-086f54ca465d)

#### Deselect the rectangle

![image](https://github.com/fabricjs/fabric.js/assets/20143684/d8c26a51-4fba-4508-8d92-246aad4bdd6a)

### Select the removed rectangle (space) : Does not work

![image](https://github.com/fabricjs/fabric.js/assets/20143684/5fbdf7d9-a57e-4c13-a1d1-379201968ad6)

## Code

```tsx

"use client"

import React, { useEffect, useRef } from "react";
import * as fabric from "fabric";

const FabricCanvas = () => {
  const canvasRef = useRef<HTMLCanvasElement>(null);

  useEffect(() => {
    const canvas = new fabric.Canvas(canvasRef?.current as HTMLCanvasElement, {
      width: 600,
      height: 400,
    });

    const rectangleSettings = {
      fill: "blue",
      width: 50,
      height: 50,
      stroke: "black",
      strokeWidth: 1,
    };

    const rows = 8;
    const columns = 10;
    const horizontalSpacing = 10;
    const verticalSpacing = 10;
    const rectangleWidth = rectangleSettings.width;
    const rectangleHeight = rectangleSettings.height;

    let x = verticalSpacing;
    let y = horizontalSpacing;

    const rectangles: fabric.Rect[] = [];

    for (let i = 0; i < rows * columns; i++) {
      const rect = new fabric.Rect(rectangleSettings);
      rect.set({ left: x, top: y });
      rectangles.push(rect);

      x += rectangleWidth + horizontalSpacing;

      if (x + rectangleWidth > canvas.width - verticalSpacing) {
        x = verticalSpacing;
        y += rectangleHeight + verticalSpacing;
      }
    }

    rectangles.forEach((rect) => {
      canvas.add(rect);
    });

    canvas.on("mouse:down", (event) => {
      const rect = event.target;
      if (rect && rect.isType('rect')) { // Check if the target is a rectangle
        if (canvas.getActiveObject() === rect) {
          canvas.remove(rect)
          // The rectangle is currently selected, implement your logic for when a rectangle is selected here.
          // Note: There's no direct `dispose` method for individual objects. Use `canvas.remove(rect)` to remove an object.
        } else {
          canvas.add(rect)
          // Set properties directly to change appearance upon selection
          rect.set({
            hasBorders: true,
            hasControls: true,
            selectable: true
          });
          canvas.setActiveObject(rect); // Explicitly set the rect as the active object
          canvas.requestRenderAll();
        }
      }})

    return () => {
      canvas.dispose();
    };
  }, []);

  return <canvas ref={canvasRef} />;
};

export default FabricCanvas;

```

**Comment by asturur** [comment, marked as answer]
There is some issue in your code, for example if a rect is the target of a mouse:down it is already in the canvas, you shouldn't add it again.
Why you just don't make the rect invisible instead of removing it?
If you remove it you have to keep the reference somewhere to add it back or you can just create a new one when you detect there wasn't any in the space where you clicked

**Reply by alamenai** [reply]
Hi @asturur ,


Yes, I made the rectangles transparent when the user clicks them:

![image](https://github.com/fabricjs/fabric.js/assets/20143684/8b075740-f4f1-47fc-a842-a26c02bdfd70)
